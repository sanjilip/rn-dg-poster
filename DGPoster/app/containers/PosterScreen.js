import React, { useEffect, useState } from 'react';
import {
    FlatList,
    View,
    Text,
    Image,
    Dimensions,
    TextInput,
    SafeAreaView,
    BackHandler,
    useColorScheme
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';

import {
    HARDWARE_BACK_PRESS,
    HEADER_BACKGROUND,
    IMAGES,
    INITIAL_PAGE,
    LEFT_ICON,
    MISSING_IMAGE,
    POSTER,
    POSTER_DATA,
    POSTER_HEADER_TITLE,
    RIGHT_ICON
} from '../utilities/constants';
import CustomHeader from '../components/CustomHeader';
import { posterScreenStyle as style } from './style';
import { API_KEYS } from '../reduxToolkit/api/apiClient/utility/apiKeys';
import { fetchPosterData } from '../reduxToolkit/api/apiClient/fetchPosterData';
import { resetPosterData } from '../reduxToolkit/slices/posterDataSlice';

const WINDOW_WIDTH = Dimensions.get('window').width;
const COLUMN_WIDTH = WINDOW_WIDTH / 3;

const PosterScreen = ({ navigation }) => {
    const COLOR_SCHEME = useColorScheme();
    const DARK_MODE = COLOR_SCHEME === 'dark';

    const dispatch = useDispatch();

    const { content, pageNumber } = useSelector(
        (state) => state.posterData ?? {}
    );

    const INITIAL_DATA = (pageNumber === INITIAL_PAGE && content) || POSTER[0];

    const [currentPage, setCurrentPage] = useState(INITIAL_PAGE);
    const [data, setData] = useState(INITIAL_DATA);
    const [searchQuery, setSearchQuery] = useState('');
    const [enableTextInput, setEnableTextInput] = useState(false);

    useEffect(() => {
        const backAction = () => {
            navigateBack();
            return true;
        };

        const backHandler = BackHandler.addEventListener(
            HARDWARE_BACK_PRESS,
            backAction
        );

        return () => backHandler.remove();
    }, []);

    useEffect(() => {
        if (content && pageNumber < 4) {
            setData(content);
        }
    }, [content]);

    const fetchNextPages = () => {
        if (!searchQuery && currentPage < 3) {
            const nextPage = currentPage + 1;
            dispatch(fetchPosterData(API_KEYS.posterData, nextPage));
            setCurrentPage(nextPage);
        }
    };

    const searchFilter = (text) => {
        /*  NOTE -  POSTER_DATA from constants is only being used for searching throughout whole data
            Lazy loading concept is handled in fetchNextPages func
            for now handling search at client end only but it should be handled at server end for smooth user experience */

        setSearchQuery(text);
        if (text) {
            const filteredData = POSTER_DATA.filter((item) =>
                item?.name.toLowerCase().includes(text.toLowerCase())
            );
            setData(filteredData);
        } else {
            dispatch(fetchPosterData(API_KEYS.posterData, INITIAL_PAGE));
            setCurrentPage(INITIAL_PAGE);
        }
    };

    const renderItem = ({ item }) => {
        const image = IMAGES[item['poster-image']] || MISSING_IMAGE;

        return (
            <View style={[style.item, { width: COLUMN_WIDTH }]}>
                <Image source={image} style={style.image} />
                <Text style={style.text}>{item.name}</Text>
            </View>
        );
    };

    const renderFlatlist = () => {
        return (
            <FlatList
                style={style.flatList}
                data={data}
                renderItem={renderItem}
                onEndReached={fetchNextPages}
                onEndReachedThreshold={0.8}
                keyExtractor={(_, index) => index.toString()}
                numColumns={3}
                showsVerticalScrollIndicator={false}
                contentContainerStyle={style.contentContainer}
            />
        );
    };

    const renderSearchInput = () => {
        return (
            enableTextInput && (
                <TextInput
                    style={[style.searchInput, DARK_MODE && { color: 'black' }]}
                    placeholder="Begin Your Search..."
                    placeholderTextColor={DARK_MODE && 'gray'}
                    value={searchQuery}
                    onChangeText={searchFilter}
                />
            )
        );
    };

    const noSearchDataFound = () => {
        const show = data.length === 0 && enableTextInput && searchQuery;
        return (
            show && (
                <Text style={style.noRecordText}>
                    Sorry! No Matching Record Found.
                </Text>
            )
        );
    };

    const navigateBack = () => {
        dispatch(resetPosterData());
        navigation?.goBack();
    };

    const onSearchIconTapped = () => {
        searchQuery &&
            dispatch(fetchPosterData(API_KEYS.posterData, INITIAL_PAGE));
        setSearchQuery('');
        setCurrentPage(INITIAL_PAGE);
        setEnableTextInput(!enableTextInput);
    };

    return (
        <SafeAreaView style={style.container}>
            <CustomHeader
                title={POSTER_HEADER_TITLE}
                leftIcon={LEFT_ICON}
                rightIcon={RIGHT_ICON}
                backgroundImage={HEADER_BACKGROUND}
                leftActionCalled={navigateBack}
                rightActionCalled={onSearchIconTapped}
            />
            {renderSearchInput()}
            {noSearchDataFound()}
            {renderFlatlist()}
        </SafeAreaView>
    );
};

export default PosterScreen;
