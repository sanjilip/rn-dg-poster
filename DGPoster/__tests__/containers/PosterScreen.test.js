import React from 'react';
import { shallow } from 'enzyme';
import { Provider, useSelector } from 'react-redux';
import configureStore from 'redux-mock-store'; // Import your Redux store configuration

import PosterScreen from '../../app/containers/PosterScreen';

jest.mock('react-native/Libraries/Utilities/useColorScheme', () => ({
    __esModule: true,
    default: jest.fn()
}));

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn(),
    useDispatch: () => jest.fn()
}));

const mockStore = configureStore([]);
const store = mockStore({
    posterData: {
        content: ['abc'],
        pageNumber: 1
    }
});

beforeEach(() => {
    useSelector.mockImplementation((callback) => {
        return callback(store);
    });
});

describe('Test PosterScreen Component', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <Provider store={store}>
                <PosterScreen navigation={{ goBack: jest.fn() }} />
            </Provider>
        );
    });

    it('renders correctly', () => {
        const wrapper = shallow(<PosterScreen />);
        expect(wrapper).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        expect(wrapper.exists()).toBe(true);
    });

    it('renders CustomHeader component', () => {
        expect(wrapper.find('CustomHeader')).toHaveLength(0);
    });

    it('renders TextInput component for search input', () => {
        expect(wrapper.find('TextInput')).toHaveLength(0);
    });

    it('renders FlatList component for posters', () => {
        expect(wrapper.find('FlatList')).toHaveLength(0);
    });
});
