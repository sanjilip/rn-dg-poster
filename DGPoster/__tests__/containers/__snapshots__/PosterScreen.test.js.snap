// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PosterScreen Component renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "black",
      "flex": 1,
    }
  }
>
  <CustomHeader
    backgroundImage={
      {
        "testUri": "../../../app/resources/images/nav_bar.png",
      }
    }
    leftActionCalled={[Function]}
    leftIcon={
      {
        "testUri": "../../../app/resources/images/back.png",
      }
    }
    rightActionCalled={[Function]}
    rightIcon={
      {
        "testUri": "../../../app/resources/images/search.png",
      }
    }
    title="Romantic Comedy"
  />
  <FlatList
    contentContainerStyle={
      {
        "flexDirection": "row",
        "flexWrap": "wrap",
      }
    }
    data={
      [
        {
          "name": "The Birds",
          "poster-image": "poster1.jpg",
        },
        {
          "name": "Rear Window",
          "poster-image": "poster2.jpg",
        },
        {
          "name": "Family Pot",
          "poster-image": "poster3.jpg",
        },
        {
          "name": "Family Pot",
          "poster-image": "poster2.jpg",
        },
        {
          "name": "Rear Window",
          "poster-image": "poster1.jpg",
        },
        {
          "name": "The Birds",
          "poster-image": "poster3.jpg",
        },
        {
          "name": "Rear Window",
          "poster-image": "poster3.jpg",
        },
        {
          "name": "The Birds",
          "poster-image": "poster2.jpg",
        },
        {
          "name": "Family Pot",
          "poster-image": "poster1.jpg",
        },
        {
          "name": "The Birds",
          "poster-image": "poster1.jpg",
        },
        {
          "name": "The Birds",
          "poster-image": "poster1.jpg",
        },
        {
          "name": "Rear Window",
          "poster-image": "poster2.jpg",
        },
        {
          "name": "Family Pot",
          "poster-image": "poster3.jpg",
        },
        {
          "name": "Family Pot",
          "poster-image": "poster2.jpg",
        },
        {
          "name": "Rear Window",
          "poster-image": "poster1.jpg",
        },
        {
          "name": "The Birds",
          "poster-image": "poster3.jpg",
        },
        {
          "name": "Rear Window",
          "poster-image": "poster3.jpg",
        },
        {
          "name": "The Birds",
          "poster-image": "poster2.jpg",
        },
        {
          "name": "Family Pot",
          "poster-image": "poster1.jpg",
        },
        {
          "name": "The Birds",
          "poster-image": "poster1.jpg",
        },
      ]
    }
    keyExtractor={[Function]}
    numColumns={3}
    onEndReached={[Function]}
    onEndReachedThreshold={0.8}
    renderItem={[Function]}
    showsVerticalScrollIndicator={false}
    style={
      {
        "backgroundColor": "black",
        "flex": 1,
        "marginHorizontal": 5,
      }
    }
  />
</RCTSafeAreaView>
`;
